### PR TITLE
SLA reporting

### PR DESCRIPTION
**Pull request by julianbrost**
This PR introduces special tables for recording data that is relevant for generating SLA reports:

* `sla_history_downtime`: One row for each downtime containing the actual start and end of the period in which the downtime was in effect.
* `sla_history_state`: One row for each hard state change.

These tables are filled in parallel to the corresponding history tables. There's also a SQL stored function giving an example implementation for computing SLA values with tests for this function in the Go code.

Having this data in tables separate from the full history tables allows this data to be kept for longer without storing the full history data for that longer period.

### Blocked by

* https://github.com/Icinga/go-libs/pull/3
* https://github.com/Icinga/icinga-testing/pull/21
* https://github.com/Icinga/docker-icingadb/pull/47

### CODE REVIEW

**Comment by lippserd** [comment]
I just wanted to let you know that the code looks good and that I'm testing it.

**Review by lippserd — CHANGES_REQUESTED** [review]
**tl;dr** Use `UNIQUE INDEX idx_sla_history_state_event(host_id, service_id, event_time)` and something similar for downtime and replace primary keys with auto-incrementing unsigned bigints.

# SLA Test

## Generator Views

```sql
CREATE OR REPLACE VIEW generator_16
AS
SELECT 0 n
UNION ALL
SELECT 1
UNION ALL
SELECT 2
UNION ALL
SELECT 3
UNION ALL
SELECT 4
UNION ALL
SELECT 5
UNION ALL
SELECT 6
UNION ALL
SELECT 7
UNION ALL
SELECT 8
UNION ALL
SELECT 9
UNION ALL
SELECT 10
UNION ALL
SELECT 11
UNION ALL
SELECT 12
UNION ALL
SELECT 13
UNION ALL
SELECT 14
UNION ALL
SELECT 15;

CREATE OR REPLACE VIEW generator_256
AS
SELECT ((hi.n << 4) | lo.n) AS n
FROM generator_16 lo,
     generator_16 hi;

CREATE OR REPLACE VIEW generator_4k
AS
SELECT ((hi.n << 8) | lo.n) AS n
FROM generator_256 lo,
     generator_16 hi;

CREATE OR REPLACE VIEW generator_64k
AS
SELECT ((hi.n << 8) | lo.n) AS n
FROM generator_256 lo,
     generator_256 hi;

CREATE OR REPLACE VIEW generator_1m
AS
SELECT ((hi.n << 16) | lo.n) AS n
FROM generator_64k lo,
     generator_16 hi;
```

## Hosts and Services

```sql
CREATE TABLE host (host_id binary(20));

INSERT INTO host
  SELECT UNHEX(SHA1(UUID())) FROM generator_4k LIMIT 1000;
  
CREATE TABLE service (host_id binary(20), service_id binary(20));

INSERT INTO service
  SELECT host_id, UNHEX(SHA1(UUID())) FROM host, (SELECT * FROM generator_16 limit 10) g;
```

## Events

```sql
DROP PROCEDURE IF EXISTS generate_random_sla_history;

DELIMITER //

CREATE PROCEDURE generate_random_sla_history(
  IN start_time int,
  IN event_interval int
)
BEGIN
  DECLARE generator_limit int;
  DECLARE zero_binary binary(20);

  SELECT FLOOR((UNIX_TIMESTAMP(NOW()) - start_time) / event_interval) INTO generator_limit;

  DROP TEMPORARY TABLE IF EXISTS host_event;
  CREATE TEMPORARY TABLE host_event (
    id int AUTO_INCREMENT,
    event_time bigint,
    hard_state tinyint,
    previous_hard_state tinyint,
    PRIMARY KEY (id)
  );

  -- All hosts share the same state events.
  INSERT INTO host_event (event_time, hard_state)
  SELECT (g.n * event_interval + start_time) * 1000, FLOOR(RAND() * 2)
  FROM (SELECT n FROM generator_1m LIMIT generator_limit) g;

  UPDATE host_event e LEFT JOIN host_event p ON p.id = e.id - 1
  SET e.previous_hard_state = COALESCE(p.hard_state, 99);

  SELECT REPEAT('\0', 20) INTO zero_binary;

  INSERT INTO sla_history_state (id, environment_id, endpoint_id, object_type, host_id, service_id, event_time,
                                 hard_state, previous_hard_state)
  SELECT UNHEX(SHA1(UUID())),
         zero_binary,
         zero_binary,
         'host',
         h.host_id,
         NULL,
         e.event_time,
         e.hard_state,
         e.previous_hard_state
  FROM host h,
       host_event e;

  DROP TEMPORARY TABLE IF EXISTS service_event;
  CREATE TEMPORARY TABLE service_event (
    id int AUTO_INCREMENT,
    event_time bigint,
    hard_state tinyint,
    previous_hard_state tinyint,
    PRIMARY KEY (id)
  );

  -- All services share the same state events.
  INSERT INTO service_event (event_time, hard_state)
  SELECT (g.n * event_interval + start_time) * 1000, FLOOR(RAND() * 4)
  FROM (SELECT n FROM generator_1m LIMIT generator_limit) g;

  UPDATE service_event e LEFT JOIN service_event p ON p.id = e.id - 1
  SET e.previous_hard_state = COALESCE(p.hard_state, 99);

  INSERT INTO sla_history_state (id, environment_id, endpoint_id, object_type, host_id, service_id, event_time,
                                 hard_state, previous_hard_state)
  SELECT UNHEX(SHA1(UUID())),
         zero_binary,
         zero_binary,
         'service',
         s.host_id,
         s.service_id,
         e.event_time,
         e.hard_state,
         e.previous_hard_state
  FROM service s,
       service_event e;

END //

DELIMITER ;

CALL generate_random_sla_history(UNIX_TIMESTAMP(CURRENT_DATE() - INTERVAL 6 MONTH), 60 * 60 * 4);
```

## Most Simple Index

Note that `object_type` has been removed.

```sql
CREATE TABLE sla_history_state_idx (
  id binary(20) NOT NULL COMMENT 'state_history.id (may reference already deleted rows)',
  environment_id binary(20) NOT NULL COMMENT 'environment.id',
  endpoint_id binary(20) DEFAULT NULL COMMENT 'endpoint.id',
  host_id binary(20) NOT NULL COMMENT 'host.id',
  service_id binary(20) DEFAULT NULL COMMENT 'service.id',

  event_time bigint unsigned NOT NULL COMMENT 'unix timestamp the event occurred',
  hard_state TINYINT UNSIGNED NOT NULL COMMENT 'hard state after this event',
  previous_hard_state TINYINT UNSIGNED NOT NULL COMMENT 'hard state before this event',

  PRIMARY KEY (id),

  UNIQUE INDEX idx_sla_history_state_event (host_id, service_id, event_time)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

INSERT INTO sla_history_state_idx SELECT id, environment_id, endpoint_id, host_id, service_id, event_time, hard_state, previous_hard_state FROM sla_history_state;
```

## Primary Key

Note that this requires `service_id` to be `NOT NULL`, so we need to insert the null binary for host events.
Also, `id` and `object_type` has been removed.

```sql
CREATE TABLE sla_history_state_pk (
  environment_id binary(20) NOT NULL COMMENT 'environment.id',
  endpoint_id binary(20) NOT NULL COMMENT 'endpoint.id',
  host_id binary(20) NOT NULL COMMENT 'host.id',
  service_id binary(20) NOT NULL COMMENT 'service.id',

  event_time bigint unsigned NOT NULL COMMENT 'unix timestamp the event occurred',
  hard_state tinyint unsigned NOT NULL COMMENT 'hard state after this event',
  previous_hard_state tinyint unsigned NOT NULL COMMENT 'hard state before this event',

  PRIMARY KEY (host_id, service_id, event_time)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;

INSERT INTO sla_history_state_pk SELECT environment_id, endpoint_id, host_id, COALESCE(service_id, REPEAT('\0', 20)), event_time, hard_state, previous_hard_state FROM sla_history_state;
```

## Comparison

### Database Configuration

Disable the query cache:

```
[mysqld]
query_cache_type=0
query_cache_size=0
```

### Buffer Pool Usage

```sql
SELECT
  COALESCE(table_name, 'All tables') AS table_name,
  COALESCE(index_name, 'All indices') AS index_name,
  pages,
  size_mib
FROM (
  SELECT
    table_name,
    index_name,
    COUNT(*)                     AS pages,
    SUM(data_size) / 1024 / 1024 AS size_mib
  FROM
    information_schema.innodb_buffer_page
  WHERE
    table_name IS NOT NULL
    AND table_name NOT LIKE '`mysql`.%'
  GROUP BY
    table_name,
    index_name
    WITH ROLLUP
) q
ORDER BY
  size_mib DESC;
```

### Random Target

```sql
SELECT
    s.host_id,
    s.service_id,
    UNIX_TIMESTAMP(CURRENT_DATE() - INTERVAL 4 MONTH) * 1000,
    UNIX_TIMESTAMP(CURRENT_DATE() - INTERVAL 2 MONTH) * 1000
INTO
    @host_id, @service_id, @start_time, @end_time
FROM
    (SELECT host_id, service_id FROM service ORDER BY RAND() LIMIT 1) s;
```

### Vanilla Table

```sql
EXPLAIN SELECT
  COUNT(*)
FROM
  sla_history_state s
WHERE
  object_type = 'service'
  AND s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time;
```

```
+------+-------------+-------+-------+-----------------------------------------------------------------+--------------------------------------+---------+------+------+------------------------------------+
| id   | select_type | table | type  | possible_keys                                                   | key                                  | key_len | ref  | rows | Extra                              |
+------+-------------+-------+-------+-----------------------------------------------------------------+--------------------------------------+---------+------+------+------------------------------------+
|    1 | SIMPLE      | s     | range | idx_sla_history_event_time,idx_sla_history_object_id_event_time | idx_sla_history_object_id_event_time | 49      | NULL |  367 | Using index condition; Using where |
+------+-------------+-------+-------+-----------------------------------------------------------------+--------------------------------------+---------+------+------+------------------------------------+
```

### Most Simple Index

```sql
EXPLAIN SELECT
  COUNT(*)
FROM
  sla_history_state_idx s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time;
```

```
+------+-------------+-------+-------+-----------------------------+-----------------------------+---------+------+------+--------------------------+
| id   | select_type | table | type  | possible_keys               | key                         | key_len | ref  | rows | Extra                    |
+------+-------------+-------+-------+-----------------------------+-----------------------------+---------+------+------+--------------------------+
|    1 | SIMPLE      | s     | range | idx_sla_history_state_event | idx_sla_history_state_event | 49      | NULL |  367 | Using where; Using index |
+------+-------------+-------+-------+-----------------------------+-----------------------------+---------+------+------+--------------------------+
```

### Primary Key

```sql
EXPLAIN SELECT
  COUNT(*)
FROM
  sla_history_state_pk s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time;
```

```
+------+-------------+-------+-------+---------------+---------+---------+------+------+--------------------------+
| id   | select_type | table | type  | possible_keys | key     | key_len | ref  | rows | Extra                    |
+------+-------------+-------+-------+---------------+---------+---------+------+------+--------------------------+
|    1 | SIMPLE      | s     | range | PRIMARY       | PRIMARY | 48      | NULL |  367 | Using where; Using index |
+------+-------------+-------+-------+---------------+---------+---------+------+------+--------------------------+
```

### First Round of Queries

```
+------------------------------+-------------+-------+-------------+
| table_name                   | index_name  | pages | size_mib    |
+------------------------------+-------------+-------+-------------+
| `sla`.`sla_history_state_pk` | PRIMARY     |  6522 | 69.04438305 |
| `sla`.`sla_history_state_pk` | All indices |  6522 | 69.04438305 |
| All tables                   | All indices |  6522 | 69.04438305 |
+------------------------------+-------------+-------+-------------+
```

```
*************************** 1. row ***************************
Query_ID: 1
Duration: 0.03820838
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state s
WHERE
  object_type = 'service'
  AND s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
*************************** 2. row ***************************
Query_ID: 2
Duration: 0.00135121
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state_idx s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
*************************** 3. row ***************************
Query_ID: 3
Duration: 0.00063988
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state_pk s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
```

### Second Round of Queries

```
+-------------------------------+--------------------------------------+-------+-------------+
| table_name                    | index_name                           | pages | size_mib    |
+-------------------------------+--------------------------------------+-------+-------------+
| All tables                    | All indices                          |  7229 | 76.26488686 |
| `sla`.`sla_history_state_pk`  | All indices                          |  6526 | 69.09804440 |
| `sla`.`sla_history_state_pk`  | PRIMARY                              |  6526 | 69.09804440 |
| `sla`.`sla_history_state`     | All indices                          |   652 |  6.49011707 |
| `sla`.`sla_history_state`     | PRIMARY                              |   623 |  6.21091843 |
| `sla`.`service`               | GEN_CLUST_INDEX                      |    45 |  0.62055969 |
| `sla`.`service`               | All indices                          |    45 |  0.62055969 |
| `sla`.`sla_history_state`     | idx_sla_history_event_time           |    23 |  0.22003841 |
| `sla`.`sla_history_state`     | idx_sla_history_object_id_event_time |     6 |  0.05916023 |
| `sla`.`sla_history_state_idx` | idx_sla_history_state_event          |     6 |  0.05616570 |
| `sla`.`sla_history_state_idx` | All indices                          |     6 |  0.05616570 |
+-------------------------------+--------------------------------------+-------+-------------+
```

```
*************************** 1. row ***************************
Query_ID: 1
Duration: 0.04388825
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state s
WHERE
  object_type = 'service'
  AND s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
*************************** 2. row ***************************
Query_ID: 2
Duration: 0.00172422
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state_idx s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
*************************** 3. row ***************************
Query_ID: 3
Duration: 0.00139034
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state_pk s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
```

### Third Round of Queries

```
+-------------------------------+--------------------------------------+-------+-------------+
| table_name                    | index_name                           | pages | size_mib    |
+-------------------------------+--------------------------------------+-------+-------------+
| All tables                    | All indices                          |  7519 | 79.04868221 |
| `sla`.`sla_history_state_pk`  | All indices                          |  6320 | 66.77701378 |
| `sla`.`sla_history_state_pk`  | PRIMARY                              |  6320 | 66.77701378 |
| `sla`.`sla_history_state`     | All indices                          |  1143 | 11.53808689 |
| `sla`.`sla_history_state`     | PRIMARY                              |  1110 | 11.20347023 |
| `sla`.`service`               | GEN_CLUST_INDEX                      |    45 |  0.62055969 |
| `sla`.`service`               | All indices                          |    45 |  0.62055969 |
| `sla`.`sla_history_state`     | idx_sla_history_event_time           |    23 |  0.22003841 |
| `sla`.`sla_history_state`     | idx_sla_history_object_id_event_time |    10 |  0.11457825 |
| `sla`.`sla_history_state_idx` | idx_sla_history_state_event          |    11 |  0.11302185 |
| `sla`.`sla_history_state_idx` | All indices                          |    11 |  0.11302185 |
+-------------------------------+--------------------------------------+-------+-------------+
```

```
*************************** 1. row ***************************
Query_ID: 1
Duration: 0.00394209
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state s
WHERE
  object_type = 'service'
  AND s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
*************************** 2. row ***************************
Query_ID: 2
Duration: 0.00063782
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state_idx s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
*************************** 3. row ***************************
Query_ID: 3
Duration: 0.00061982
   Query: SELECT
  COUNT(*)
FROM
  sla_history_state_pk s
WHERE
  s.host_id = @host_id
  AND s.service_id = @service_id
  AND s.event_time BETWEEN @start_time AND @end_time
ORDER BY event_time
```

## Conclusion

The unique index works just as well as the primary key when index parts have been loaded into the buffer pool. And since in InnoDB each record in a secondary index also contains the primary key, it makes sense to replace the binary with an auto-incrementing unsigned bigint. The same should be applied to the downtime table.

**Comment by julianbrost** [comment]
> **tl;dr** Use `UNIQUE INDEX idx_sla_history_state_event(host_id, service_id, event_time)` and something similar for downtime and replace primary keys with auto-incrementing unsigned bigints.

Well, state was the easy case, downtimes have two event timestamps per row and are not uniquely identified by them, so not even `(host_id, service_id, downtime_start, downtime_end)` would be a valid unique key there, you need `downtime_id`.

File: pkg/types/state_type.go [hunk 1]
```diff
@@ -46,10 +46,15 @@ func badStateType(t interface{}) error {
 	return errors.Errorf("bad state type: %#v", t)
 }
 
+var (
```
Comment (Al2Klimov) [inline pkg/types/state_type.go:49]: "var"?

**Comment by julianbrost** [comment]
> replace primary keys with auto-incrementing unsigned bigints

Skipped that part as discussed earlier today as for PostgreSQL, this would require some extra special-casing in the code as there you have to name the constraint which triggers the update if it is violated.

I also looked into possible indices for PostgreSQL and in theory, the following would be quite nice:

```
CREATE INDEX testidx ON sla_history_downtime USING gist (int8range(downtime_start, downtime_end));
```

However:
* There seems to be no easy way to include `host_id` and `service_id` so that the index is partitioned by checkable so to say. This could probably be worked around, but I guess it's not worth the effort for now.
  * As GiST is quite extensible, it should be possible to build some custom type that encapsulates the checkable ID and the range with custom operators and specialized index operations.
  * Some messing around with using the `box` type (representing a rectangle in a 2D plane) where one dimension represents time and the other one the checkable ID space, but that would be super ugly in the schema and queries.
* And I haven't looked at insert performance so far, but there are [hints in the documentation](https://www.postgresql.org/docs/14/gist-implementation.html) that this is something that should be considered.

**Comment by julianbrost** [comment]
Anyways, feel free to have another look at this. I expect there to only be changes in the following areas:
* Tests: try to make use of icinga-testing here so that "start database, wait for it and import schema" doesn't have to be implemented in the workflow in this repo again.
* MySQL function: probably adapt to some of the changes that were required for PostgreSQL so that both functions are more similar: change input variable names and sort by an explicit column rather than by a case statement.
* PostgreSQL function: decide whether to keep using `int8range` with its operators in the `WHERE` clauses and maybe restructure the huge `UNION` query to use `WITH`.

**Comment by julianbrost** [comment]
Should be ready now apart from the TODOs in the description, so might be a good idea to start reviewing the dependencies first.

**Review by lippserd — CHANGES_REQUESTED** [review]
The `object_type` conditions can also be safely removed from the WHERE clauses.

File: schema/pgsql/schema.sql [hunk 2]
```diff
@@ -21,6 +21,143 @@ CREATE TYPE comment_type AS ENUM ( 'comment', 'ack' );
...
+    IF cur.event_type = 'state_change' THEN
+      last_hard_state := cur.hard_state;
+    ELSEIF cur.event_type = 'downtime_start' THEN
+      active_downtimes := active_downtimes + 1;
+    ELSEIF cur.event_type = 'downtime_end' THEN
+      active_downtimes := active_downtimes - 1;
+    END IF;
+  END LOOP;
+
+  total_time := in_end_time - in_start_time;
```
[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/schema.sql:155]: What value would you expect if in your example, the host recovered in the middle of that 1 hour period (i.e. did not exist for almost a year, down for 30 minutes, up for 30 minutes)?

I see two approaches here:
1. Special-casing that if the state at the very beginning was UNKNOWN/99, use the first known state, so in above example, everything except 30 minutes would count as down, so value would be something like 0.x%.
2. Try to detect that the available data range is shorter then the requested range and truncate it accordingly, so 50% in the above example.

Number 2 would lead to different results between creating an object (thus seeing a UNKNOWN/99 state in the database so we know it was just created) and only enabling Icinga DB within the requested interval (data is just cut off and some value would be assumed for everything before).

[the remark this answers is not in the document]
Comment (lippserd) [reply schema/pgsql/schema.sql:155]: Data can only be truncated when removed from the stream, right? This means that Icinga DB is not running for a long time. I don't think we can solve that either.

File: schema/pgsql/schema.sql [hunk 3]
```diff
@@ -21,6 +21,143 @@ CREATE TYPE comment_type AS ENUM ( 'comment', 'ack' );
 CREATE TYPE notification_type AS ENUM ( 'downtime_start', 'downtime_end', 'downtime_removed', 'custom', 'acknowledgement', 'problem', 'recovery', 'flapping_start', 'flapping_end' );
 CREATE TYPE history_type AS ENUM ( 'notification', 'state_change', 'downtime_start', 'downtime_end', 'comment_add', 'comment_remove', 'flapping_start', 'flapping_end', 'ack_set', 'ack_clear' );
 
+CREATE OR REPLACE FUNCTION reports_get_sla_ok_percent(
```
[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/schema.sql:24]: How did you conclude that PARALLEL RESTRICTED is the correct choice? Documentation seems a bit lacking only listing examples for things you can't do. Does `FOR row IN query LOOP` qualify as accessing cursors?

Anyways, seems like `PARALLEL *` was only added in 9.6 and CentOS/RHEL 7 is still on 9.5 :upside_down_face: from a quick search there's also nothing similar to `/*!` in MySQL.

File: schema/pgsql/schema.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/schema.sql:155]: > I would go with `MAX(in_start_time, time_of_first_event)`.

Really for any event, not only if the previous state was UNKNOWN/99? This would then require having one event from before the report interval (or exactly at the beginning) so complicating retention/cleanup.

> However, we would still have ambiguities with decommissioned/recommissioned objects during the reporting period. But I don't think we can solve this without additional events.

I think excluding any time in a UNKNOWN/99 state should also produce acceptable results there. When something is recommissioned, it should start in that state again and the time it did not exist could be excluded from the total.

[the remark this answers is not in the document]
Comment (lippserd) [reply schema/pgsql/schema.sql:155]: > Really for any event, not only if the previous state was UNKNOWN/99? This would then require having one event from before the report interval (or exactly at the beginning) so complicating retention/cleanup.

We already select the initial state. I would also select the time of the initial state and use that for the `MAX(...`. There shouldn't be any downsides, right?

> I think excluding any time in a UNKNOWN/99 state should also produce acceptable results there. When something is recommissioned, it should start in that state again and the time it did not exist could be excluded from the total.

The problem is that you don't know when it was taken out of service, only when it was put back into service. So I wouldn't try to guess anything here as it can only be wrong.

[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/schema.sql:155]: > Data can only be truncated when removed from the stream, right? This means that Icinga DB is not running for a long time. I don't think we can solve that either.

Data can be missing in the SQL database under the following circumstances:
1. At the beginning:
  a) You start using Icinga DB without migrating history from IDO.
  b) You are using the cleanup/retention feature.
  Both will work but there will be incorrect results for reports including earlier times.
2. At the end:
  a) Some of the latest history items is still processed at the moment. That should be no more than a few seconds though, so even if it's missing, it won't have a huge impact on the result.
  b) Icinga DB is currently not running. Then that interval become bigger, data is still cached in Redis though.
  All will be fine if things are running properly when generating the report.
3. Somewhere in the middle:
  This should only happen if you do things like stop Redis and then restart Icinga 2 while Redis is not running, so it will discard the pending history. That data is lost and there isn't anything we can do about this here.

[the remark this answers is not in the document]
Comment (julianbrost) [reply schema/pgsql/schema.sql:155]: > We already select the initial state. I would also select the time of the initial state and use that for the `MAX(...`. There shouldn't be any downsides, right?

Then you can no longer just use `time < now-1y` to delete from the SLA history tables, otherwise when doing reports that start shortly after the beginning of the available data, you might reduce the total time for many hosts (it gets worse the more stable they are). I think that was the reason why we added `previous_hard_state` in the first place, so we can rely only on events within the report interval.

**Comment by julianbrost** [comment]
@lippserd Please have a look at the changes done so far (I did only change PostgreSQL for now, once we're happy with that I'll adapt MySQL). I implemented some basic handling for UNKNOWN/99, for now it doesn't try to handle things like downtimes within an UNKNOWN interval.

For example, consider the following sequence of events:

1. critical
2. downtime start
3. downtime end
4. unprovision
5. reprovision
6. unknown -> ok

With the current code, the time from 3 to 6 would be considered UNKNOWN and excluded from the total time, whereas if the downtime wasn't there, it would be 1 to 6. From the events we can't also tell for sure if the downtime or the reprovisioning happened first. But with this kind of situation, we're down to guessing already anyways, just imagine scheduling a long-lasting downtime and then reprovisioning the host, downtime will be removed, but no cancel event will be written to the history.

The code could also be adapted to exclude the time from the last state event, but that would make the SQL functions more complex (would need some extra `pending_problem_time` variable counting the problem time from the last state event and the next state event either committing that to `problem_time` or discarding it and reducing `total_time` instead). I mean that's doable for sure, but still added complexity for questionable gain, so I'd rather keep this as implemented in the current draft state.

**Comment by lippserd** [comment]
> I mean that's doable for sure, but still added complexity for questionable gain, so I'd rather keep this as implemented in the current draft state.

Just for the record, since we've had this topic in a discussion before: Sounds good to me too!

Changes are looking good so far. I will test them!

**Comment by julianbrost** [comment]
A few more changes will be coming, see the last 4 items I added to the todo list in the PR description.

**Comment by julianbrost** [comment]
Btw, regarding the function name: shall we drop the `reports_` Prefix? Looks a bit pointless and is basically there because I took the IDO SLA stuff as a starting point. But probably the prefix is only there as the reports module creates this function in the IDO database and that's not the case here.

**Comment by lippserd** [comment]
> Btw, regarding the function name: shall we drop the `reports_` Prefix? Looks a bit pointless and is basically there because I took the IDO SLA stuff as a starting point. But probably the prefix is only there as the reports module creates this function in the IDO database and that's not the case here.

Yeah, let's drop it!

**Comment by julianbrost** [comment]
Should be mostly there now.

One thing I'm not fully happy with is naming ("history-retention" vs. just "retention") and things are still a bit inconsistent there. I wanted to rename everything to just "retention" so that you can then use history retention and SLA retention to separate both, but then while reworking it, I figured that "history retention" might be easier to understand than just "retention" and the SLA history also somewhat qualifies as history.

**Comment by julianbrost** [comment]
> One thing I'm not fully happy with is naming ("history-retention" vs. just "retention") and things are still a bit inconsistent there. I wanted to rename everything to just "retention" so that you can then use history retention and SLA retention to separate both, but then while reworking it, I figured that "history retention" might be easier to understand than just "retention" and the SLA history also somewhat qualifies as history.

I went with just "retention" now. Rationale: in the configuration, you will have the sub-key "history-retention" so things should be pretty clear there. Other place where this might be visible are log messages, but as retention only logs debug messages anyways, There's a pretty good chance you should know what you're looking at when you see it.

**Comment by julianbrost** [comment]
Rebased to resolve merge conflict in `1.0.0.sql`. Tiny other change: PR no longer changes the capitalization of `TIMESTAMP` in the insert into the schema version table.

File: pkg/icingadb/v1/history/downtime.go [hunk 4]
```diff
@@ -109,6 +132,18 @@ func (et DowntimeEventTime) Value() (driver.Value, error) {
 	}
 }
 
+type SlaDowntimeEndTime struct {
+	History *SlaHistoryDowntime `db:"-"`
+}
+
+func (et SlaDowntimeEndTime) Value() (driver.Value, error) {
```
Comment (lippserd) [inline pkg/icingadb/v1/history/downtime.go:139]: I would add the standard comment for implementing an interface here.

**Review by lippserd — APPROVED** [review]
Added a note about a missing function doc, but that's not a problem at all, so feel free to merge. Good work!

**Comment by julianbrost** [comment]
So that was the version bump for https://github.com/Icinga/icinga-testing/pull/21 being merged. I'll add a comment in a sec (https://github.com/Icinga/icingadb/pull/247#discussion_r872201109).

**Review by lippserd — APPROVED** [review]
Thanks for the two comments!
